lib/AST/RequirementMachine/RequirementLowering.cpp

```diff
@@ -455,29 +526,136 @@ void swift::rewriting::realizeInheritedRequirements(
+                         subjectType, constraintString);
+      diagnosedError = true;
+
+      auto getNameWithoutSelf = [&](std::string subjectTypeName) {
+        std::string selfSubstring = "Self.";
+
+        if (subjectTypeName.rfind(selfSubstring, 0) == 0) {
+          return subjectTypeName.erase(0, selfSubstring.length());
+        }
+
+        return subjectTypeName;
+      };
+
+      if (allowConcreteGenericParams) {
+        auto subjectTypeName = subjectType.getString();
+        auto subjectTypeNameWithoutSelf = getNameWithoutSelf(subjectTypeName);
+        ctx.Diags.diagnose(loc, diag::requires_conformance_nonprotocol_fixit,
+                           subjectTypeNameWithoutSelf, constraintString)
+             .fixItReplace(loc, " == ");
+      }
+
+      break;
+    }
+
+    case RequirementError::Kind::ConcreteTypeMismatch: {
+      auto type1 = error.requirement.getFirstType();
+      auto type2 = error.requirement.getSecondType();
+
+      if (!type1->hasError() && !type2->hasError()) {
+        ctx.Diags.diagnose(loc, diag::requires_same_concrete_type,
+                           type1, type2);
+        diagnosedError = true;
+      }
+
+      break;
+    }
+
+    case RequirementError::Kind::NonTypeParameter: {
+      ctx.Diags.diagnose(loc, diag::requires_not_suitable_archetype,
+                         error.requirement.getFirstType());
+      diagnosedError = true;
+      break;
+    }
+
+    case RequirementError::Kind::RedundantRequirement: {
+      auto requirement = error.requirement;
+      switch (requirement.getKind()) {
+      case RequirementKind::SameType:
+        ctx.Diags.diagnose(loc, diag::redundant_same_type_to_concrete,
+                           requirement.getFirstType(),
+                           requirement.getSecondType());
+        break;
+      case RequirementKind::Conformance:
+        ctx.Diags.diagnose(loc, diag::redundant_conformance_constraint,
+                           requirement.getFirstType(),
+                           requirement.getProtocolDecl());
+        break;
+      case RequirementKind::Superclass:
+        ctx.Diags.diagnose(loc, diag::redundant_superclass_constraint,
+                           requirement.getFirstType(),
+                           requirement.getSecondType());
+        break;
+      case RequirementKind::Layout:
+        ctx.Diags.diagnose(loc, diag::redundant_layout_constraint,
+                           requirement.getFirstType(),
+                           requirement.getLayoutConstraint());
+        break;
+      }
+
+      break;
+    }
+    }
+  }
+
+  return diagnosedError;
+}
+
 ArrayRef<StructuralRequirement>
 StructuralRequirementsRequest::evaluate(Evaluator &evaluator,
                                         ProtocolDecl *proto) const {
   assert(!proto->hasLazyRequirementSignature());
 
   SmallVector<StructuralRequirement, 4> result;
+  SmallVector<RequirementError, 4> errors;
```

Comment:
You can diagnose these here too -- this request runs for protocols defined in source